File: network_secret/juniper8.py
"""
Encrypt and decrypt Juniper $8$ (type 8) passwords.

Unlike the reversible, keyless $9$ substitution cipher, the $8$ format is
genuine authenticated encryption keyed by the device master password
(``set system master-password``). The same master password is required to
both encrypt and decrypt: without it, $8$ secrets cannot be recovered.

Juniper documents the format, but the documentation is incomplete: it omits
that only the first 12 bytes of the 16-byte iv field are used as the GCM
nonce, so a by-the-book implementation fails authentication. That missing
detail was reverse-engineered and verified against a real JUNOS 23.2 device
(tag authentication passes):

    $8$<crypt-algo>$<hash-algo>$<iterations>$<salt>$<iv>$<tag>$<ciphertext>

* encoding : standard base64 (RFC 4648), no padding, for every binary field.
* key      : PBKDF2-HMAC-SHA256(master_password, salt, iterations) -> 32 bytes.
* cipher   : AES-256-GCM, no additional authenticated data (AAD).
* nonce    : the iv field decodes to 16 bytes, but only the first 12 are used
             as the GCM nonce; the trailing 4 bytes are unused padding.

Public API:
    decrypt(ciphertext: str, master_password: str) -> str
    encrypt(plaintext: str, master_password: str) -> str
    check(ciphertext: str, other: str, master_password: str) -> tuple[str, str, bool]
"""

from __future__ import annotations

import base64
import binascii
import os

from cryptography.exceptions import InvalidTag
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.hashes import SHA256
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC

__all__ = ["decrypt", "encrypt", "check"]

MAGIC = "$8$"

# Environment variable the CLI reads the master password from when --master
# is not given on the command line.
ENV_MASTER = "JUNOS_MASTER_PASSWORD"

# Fixed algorithm parameters JUNOS currently emits for type 8.
CRYPT_ALGO = "aes256-gcm"
HASH_ALGO = "hmac-sha2-256"
DEFAULT_ITERATIONS = 100
# Bounds JUNOS accepts for the type 8 iteration count. Enforced on decrypt so a
# hostile $8$ value cannot force an arbitrarily expensive PBKDF2 derivation.
MIN_ITERATIONS = 10
MAX_ITERATIONS = 10000
SALT_LEN = 8  # bytes
IV_LEN = 16  # bytes stored in the string
NONCE_LEN = 12  # bytes of the IV actually used as the GCM nonce
KEY_LEN = 32  # bytes, AES-256
TAG_LEN = 16  # bytes, GCM authentication tag


def _b64encode(data: bytes) -> str:
    return base64.b64encode(data).decode("ascii").rstrip("=")


def _b64decode(text: str) -> bytes:
    try:
        return base64.b64decode(text + "=" * (-len(text) % 4), validate=True)
    except binascii.Error as e:
        raise ValueError(f"Invalid base64 in $8$ value: {text!r}") from e


def _derive_key(master_password: str, salt: bytes, iterations: int) -> bytes:
    kdf = PBKDF2HMAC(
        algorithm=SHA256(), length=KEY_LEN, salt=salt, iterations=iterations
    )
    return kdf.derive(master_password.encode("utf-8"))


def encrypt(plaintext: str, master_password: str) -> str:
    """Encrypt plaintext into a Juniper $8$ value under master_password.

    Output is non-deterministic: a fresh random salt and IV are generated on
    every call, so the same plaintext produces a different $8$ string each
    time. All of them decrypt back to the same plaintext with the same master
    password.
    """
    salt = os.urandom(SALT_LEN)
    iv = os.urandom(IV_LEN)
    key = _derive_key(master_password, salt, DEFAULT_ITERATIONS)

    # JUNOS uses only the first 12 IV bytes as the GCM nonce. AESGCM appends
    # the authentication tag to the ciphertext.
    sealed = AESGCM(key).encrypt(iv[:NONCE_LEN], plaintext.encode("utf-8"), None)
    ciphertext, tag = sealed[:-TAG_LEN], sealed[-TAG_LEN:]

    return MAGIC + "$".join(
        [
            CRYPT_ALGO,
            HASH_ALGO,
            str(DEFAULT_ITERATIONS),
            _b64encode(salt),
            _b64encode(iv),
            _b64encode(tag),
            _b64encode(ciphertext),
        ]
    )


def decrypt(ciphertext: str, master_password: str) -> str:
    """Decrypt a Juniper $8$ value with master_password.

    Raises ValueError on a malformed $8$ string or on authentication failure
    (wrong master password, or a value not produced by this scheme).
    """
    if not ciphertext.startswith(MAGIC):
        raise ValueError(f"Not a Juniper $8$ string: must start with {MAGIC}")

    # Leading '$' yields an empty first element; '8' is the type tag.
    parts = ciphertext.split("$")
    if len(parts) != 9:
        raise ValueError("Malformed $8$ value: expected 7 fields after the prefix")
    _, _type, crypt_algo, hash_algo, iters, salt_b, iv_b, tag_b, ct_b = parts

    if crypt_algo != CRYPT_ALGO:
        raise ValueError(f"Unsupported crypt-algo {crypt_algo!r}: only {CRYPT_ALGO}")
    if hash_algo != HASH_ALGO:
        raise ValueError(f"Unsupported hash-algo {hash_algo!r}: only {HASH_ALGO}")
    if not (iters.isascii() and iters.isdigit()):
        # str.isdigit() is True for non-ASCII digits (e.g. superscript "²")
        # that int() cannot parse; require plain ASCII digits.
        raise ValueError(f"Invalid iteration count: {iters!r}")
    if not MIN_ITERATIONS <= int(iters) <= MAX_ITERATIONS:
        raise ValueError(
            f"Iteration count {iters} out of range "
            f"({MIN_ITERATIONS}-{MAX_ITERATIONS})"
        )

    salt, iv, tag, ct = (
        _b64decode(salt_b),
        _b64decode(iv_b),
        _b64decode(tag_b),
        _b64decode(ct_b),
    )
    key = _derive_key(master_password, salt, int(iters))
    try:
        plaintext = AESGCM(key).decrypt(iv[:NONCE_LEN], ct + tag, None)
    except InvalidTag as e:
        raise ValueError(
            "Authentication failed: wrong master password, or the value was "
            "not produced by this scheme"
        ) from e
    return plaintext.decode("utf-8")


def check(
    ciphertext: str, other: str, master_password: str
) -> tuple[str, str, bool]:
    """Decrypt `ciphertext` and compare to `other`.

    `other` may be a plaintext, or another $8$ value (auto-detected by the
    $8$ prefix and decrypted with the same master password). Returns
    (plain_a, plain_b, match).
    """
    plain_a = decrypt(ciphertext, master_password)
    plain_b = (
        decrypt(other, master_password) if other.startswith(MAGIC) else other
    )
    return plain_a, plain_b, plain_a == plain_b

File: web/src/lib/ciphers/juniper8.ts
/**
 * Juniper/HPE $8$ (type 8) authenticated encryption.
 *
 * Unlike $9$, $8$ is genuine authenticated encryption keyed by the device
 * master password. Reverse-engineered and verified against a real JUNOS 23.2
 * device:
 *
 *   $8$<crypt-algo>$<hash-algo>$<iterations>$<salt>$<iv>$<tag>$<ciphertext>
 *
 * - encoding : standard base64, no padding, for every binary field.
 * - key      : PBKDF2-HMAC-SHA256(master, salt, iterations) -> 32-byte AES key.
 * - cipher   : AES-256-GCM, no additional authenticated data.
 * - nonce    : the iv field is 16 bytes but only the first 12 are the GCM
 *              nonce; the trailing 4 bytes are unused.
 *
 * All crypto runs in the browser via the Web Crypto API (SubtleCrypto).
 */

import type { Cipher } from "./types";
import { base64ToBytes, bytesToBase64 } from "../base64";

const MAGIC = "$8$";
const CRYPT_ALGO = "aes256-gcm";
const HASH_ALGO = "hmac-sha2-256";
const DEFAULT_ITERATIONS = 100;
const SALT_LEN = 8;
const IV_LEN = 16;
const NONCE_LEN = 12;
const TAG_LEN = 16;

function getCrypto(): Crypto {
  const c = globalThis.crypto;
  if (!c || !c.subtle) {
    throw new Error("Web Crypto API is not available in this environment");
  }
  return c;
}

/**
 * Cast a Uint8Array to BufferSource. TypeScript's DOM lib (>=5.7) narrows
 * BufferSource to ArrayBuffer-backed views, which our plain Uint8Arrays
 * satisfy at runtime but not in the type system.
 */
function buf(bytes: Uint8Array): BufferSource {
  return bytes as unknown as BufferSource;
}

async function deriveKey(
  master: string,
  salt: Uint8Array,
  iterations: number,
): Promise<CryptoKey> {
  const c = getCrypto();
  const base = await c.subtle.importKey(
    "raw",
    buf(new TextEncoder().encode(master)),
    "PBKDF2",
    false,
    ["deriveKey"],
  );
  return c.subtle.deriveKey(
    { name: "PBKDF2", salt: buf(salt), iterations, hash: "SHA-256" },
    base,
    { name: "AES-GCM", length: 256 },
    false,
    ["encrypt", "decrypt"],
  );
}

export async function encode(plaintext: string, master?: string): Promise<string> {
  if (!master) {
    throw new Error("A master password is required to encrypt a $8$ value");
  }
  const c = getCrypto();
  const salt = c.getRandomValues(new Uint8Array(SALT_LEN));
  const iv = c.getRandomValues(new Uint8Array(IV_LEN));
  const key = await deriveKey(master, salt, DEFAULT_ITERATIONS);
  const nonce = iv.slice(0, NONCE_LEN);
  const sealed = new Uint8Array(
    await c.subtle.encrypt(
      { name: "AES-GCM", iv: buf(nonce), tagLength: 128 },
      key,
      buf(new TextEncoder().encode(plaintext)),
    ),
  );
  const ciphertext = sealed.slice(0, sealed.length - TAG_LEN);
  const tag = sealed.slice(sealed.length - TAG_LEN);
  return (
    MAGIC +
    [
      CRYPT_ALGO,
      HASH_ALGO,
      String(DEFAULT_ITERATIONS),
      bytesToBase64(salt),
      bytesToBase64(iv),
      bytesToBase64(tag),
      bytesToBase64(ciphertext),
    ].join("$")
  );
}

export async function decode(encoded: string, master?: string): Promise<string> {
  if (!master) {
    throw new Error("A master password is required to decrypt a $8$ value");
  }
  encoded = encoded.trim();
  if (!encoded.startsWith(MAGIC)) {
    throw new Error("Not a Juniper/HPE $8$ string: must start with $8$");
  }
  const parts = encoded.split("$");
  if (parts.length !== 9) {
    throw new Error("Malformed $8$ value: expected 7 fields after the prefix");
  }
  const [, , cryptAlgo, hashAlgo, iters, saltB, ivB, tagB, ctB] = parts;
  if (cryptAlgo !== CRYPT_ALGO) {
    throw new Error(`Unsupported crypt-algo '${cryptAlgo}': only ${CRYPT_ALGO}`);
  }
  if (hashAlgo !== HASH_ALGO) {
    throw new Error(`Unsupported hash-algo '${hashAlgo}': only ${HASH_ALGO}`);
  }
  if (!/^\d+$/.test(iters)) {
    throw new Error(`Invalid iteration count: '${iters}'`);
  }
  const salt = base64ToBytes(saltB);
  const iv = base64ToBytes(ivB);
  const tag = base64ToBytes(tagB);
  const ciphertext = base64ToBytes(ctB);

  const key = await deriveKey(master, salt, parseInt(iters, 10));
  const data = new Uint8Array(ciphertext.length + tag.length);
  data.set(ciphertext);
  data.set(tag, ciphertext.length);
  const nonce = iv.slice(0, NONCE_LEN);
  try {
    const c = getCrypto();
    const plaintext = await c.subtle.decrypt(
      { name: "AES-GCM", iv: buf(nonce), tagLength: 128 },
      key,
      buf(data),
    );
    return new TextDecoder().decode(plaintext);
  } catch {
    throw new Error(
      "Authentication failed: wrong master password, or the value was not produced by this scheme",
    );
  }
}

export const juniper8: Cipher = {
  id: "juniper8",
  name: "Juniper/HPE $8$",
  vendor: "Juniper/HPE",
  magic: "$8$",
  tagline:
    "Juniper/HPE type 8 ($8$): master-password-keyed AES-256-GCM encryption for JUNOS secrets.",
  reversible: false,
  keyed: true,
  keyLabel: "Master password",
  status: "available",
  notes: [
    "$8$ is real authenticated encryption: PBKDF2-HMAC-SHA256 derives an AES-256-GCM key from the master password.",
    "Decryption requires the same master password set on the device (set system master-password).",
    "The 16-byte iv field is stored in full, but only its first 12 bytes are used as the GCM nonce.",
    "A wrong master password fails the GCM authentication tag, so decoding reports an error instead of garbage.",
  ],
  example: {
    encoded:
      "$8$aes256-gcm$hmac-sha2-256$100$p8XEvHtxRNE$d/hqRmh5etkBzo7WSdtvjg$7w1eMTYXkz4RdzMF9CAkJQ$qVLunbFwBWwyxln2Vg",
    plaintext: "LabBgpSecret1",
    key: "a3f8d9e112c04b7af1c3e8b92d057a4e",
  },
  links: [
    {
      label: "github.com/antoinekh/network-secret",
      url: "https://github.com/antoinekh/network-secret",
      note: "Python library & CLI, plus how the format was reverse-engineered.",
    },
  ],
  encode,
  decode,
};
